portable/GCC/ARM_CRx_MPU/portASM.S

```diff
@@ -0,0 +1,580 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://github.com/FreeRTOS
+ *
+ */
+
+    .arm
+    .syntax unified
+    /* All code in the portASM.S file is intended to be run from a prvileged
+     * operating mode, as such mark the entire file as privileged_functions */
+    .section privileged_functions
+
+#define FREERTOS_ASSEMBLY
+    #include "portmacro_asm.h"
+    #include "mpu_syscall_numbers.h"
+#undef FREERTOS_ASSEMBLY
+
+    /* External FreeRTOS-Kernel Variables */
+    .extern pxCurrentTCB
+    .extern uxSystemCallImplementations
+    .extern ulPortInterruptNesting
+    .extern ulICCEOIR
+    .extern ulPortYieldRequired
+    .extern __privileged_functions_start__
+    .extern __privileged_functions_end__
+    .extern __privileged_stacks_start__
+    .extern __privileged_stacks_end__
+    .extern __syscalls_flash_length__
+    .extern __syscalls_flash_start__
+    .extern __syscalls_flash_end__
+
+    /* External FreeRTOS-Kernel Functions */
+    .extern vAssertCalled
+    .extern vTaskSwitchContext
+    .extern vApplicationIRQHandler
+
+/* ----------------------------------------------------------------------------------- */
+/* Save the register context of a FreeRTOS Task. */
+.macro portSAVE_CONTEXT
+    DSB
+    ISB
+    /* Push R0 and the Link Register (LR) for scratch register space */
+    PUSH    { R0, LR }
+    /* Load the pointer to the current task's Task Control Block (TCB) */
+    LDR     LR, =pxCurrentTCB
+    /* Load the actual TCB into LR */
+    LDR     LR, [LR]
+    /* Set LR to pxTopOfStack, the address of where to save the task context */
+    LDR     LR, [LR]
+
+    /* Load the address of ulCriticalNesting */
+    LDR     R0, =ulCriticalNesting
+    /* Load the value of ulCriticalNesting into R0 */
+    LDR     R0, [R0]
+    /* Push the value of ulCriticalNesting into the context */
+    STM     LR!, {R0}
+
+#if ( portENABLE_FPU == 1 )
+    /* Save the floating point context */
+    /* Copy the Floating Point Status and Control Register (FPSRC) */
+    FMRX    R0, FPSCR
+    /* Push the value of FPSCR onto the stack */
+    STM     LR!, { R0 }
+    /* Push the 32 Floating Point Registers (FPRs) onto the stack */
+    VSTM    LR!, { D0-D15 }
+#endif /* ( portENABLE_FPU == 1 ) */
+
+    /* Restore the saved register */
+    POP     { R0 }
+    /* Save the pre-exception Registers, Stack Pointer (SP), and LR */
+    STM     LR, { R0-R14 }^
+    /* Increment the LR after the popped registers */
+    ADD     LR, LR, #portGPR_LENGTH
```

Comment:
Can you expand the comment here to explain why this is required? (I think its touched upon in functions further below).

Reply:
Comment adjusted to be more clear